gateway/src/main/java/io/camunda/zeebe/gateway/admin/backup/BackupRequestHandler.java

```diff
@@ -25,40 +31,144 @@ public BackupRequestHandler(final BrokerClient brokerClient) {
+            .map(partitionId -> getBackupRequestForPartition(backupId, partitionId))
             .map(brokerClient::sendRequestWithRetry)
             .toArray(CompletableFuture[]::new);
 
     return CompletableFuture.allOf(backupTriggered)
         .thenApply(ignore -> backupId)
         .exceptionallyCompose(
-            error -> CompletableFuture.failedFuture(backupFailed(backupId, error.getCause())));
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("take", backupId, error.getCause())));
+  }
+
+  @Override
+  public CompletableFuture<BackupStatus> getStatus(final long backupId) {
+    final Either<Throwable, Boolean> topologyComplete = checkTopologyComplete();
+    if (topologyComplete.isLeft()) {
+      return CompletableFuture.failedFuture(
+          operationFailed("query", backupId, topologyComplete.getLeft()));
+    }
+
+    final var statusesReceived =
+        topologyManager.getTopology().getPartitions().stream()
+            .map(partitionId -> getStatusQueryForPartition(backupId, partitionId))
+            .map(brokerClient::sendRequestWithRetry)
+            .toList();
+
+    return CompletableFuture.allOf(statusesReceived.toArray(CompletableFuture[]::new))
+        .thenApply(ignore -> aggregatePartitionStatus(backupId, statusesReceived))
+        .exceptionallyCompose(
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("query", backupId, error.getCause())));
+  }
+
+  private Either<Throwable, Boolean> checkTopologyComplete() {
+    final BrokerClusterState topology = topologyManager.getTopology();
+    if (topology == null) {
+      return Either.left(new NoTopologyAvailableException());
+    }
+
+    final int expectedPartitionCount = topology.getPartitionsCount();
+    final int knownPartitions = topology.getPartitions().size();
+    if (expectedPartitionCount != knownPartitions) {
+      return Either.left(
+          new IncompleteTopologyException(
+              "Expected to send request to all %d partitions, but found only %d partitions in topology."
+                  .formatted(expectedPartitionCount, knownPartitions)));
+    }
+
+    return Either.right(true);
+  }
+
+  private static BackupOperationFailedException operationFailed(
+      final String operation, final long backupId, final Throwable error) {
+    return new BackupOperationFailedException(operation, backupId, error);
+  }
+
+  private BackupStatus aggregatePartitionStatus(
+      final long backupId,
+      final List<CompletableFuture<BrokerResponse<BackupStatusResponse>>> completedFutures) {
+
+    final var partitionStatuses =
+        completedFutures.stream()
+            .map(response -> response.join().getResponse())
+            .map(PartitionBackupStatus::from)
+            .toList();
+
+    final var combinedStatus = getAggregatedStatus(partitionStatuses);
+
+    if (combinedStatus.isEmpty()) {
```

Comment:
:question: If I understood correctly, this can only happen if we did not send any requests, correct? If that's the case, the error message is somewhat obscure. Is there another case where this could be empty but we did send some requests? Or is this for future proofing?

Reply:
This would only happen if no response list is empty. This should never happen because if there are no partitions in the topology, it should fail before reaching this step. This check is added just as a safety condition.

Reply:
:x: Can we reflect that in the error message?

```java
throw new IncompleteTopologyException("Backup status cannot be calculated from %s, as the topology is incomplete".formatted(partitionStatuses))
```

I think it will help us faster understand what's going on in case anything goes wrong.

Reply:
:+1: I have updated the error message. But kept the exception as `IllegalStateException` because this should never happen.